execution_invariance_test/test.py

```diff
@@ -0,0 +1,154 @@
+        module_name = code_file.stem
+        if module_name in sys.modules:
+            print(f"Deleting module {module_name}")
+            del sys.modules[module_name]
+
+
+def run_tests(projects, project_module_map):
+    invariance_test_node_id = "tests/stack_test.py::test_stack"
+    failed = False
+    # Run the tests with and without instrumentation
+    for project in projects:
+        print(f"Installing project {project.name}")
+        install_project(project)
+        print(f"Running project {project.name}")
+        test_result = run_project(project)
+        remove_related_modules(project, project_module_map[project.name])
+        instrumented_test_result = run_instrumented_project(project)
+
+        for test, result in test_result.items():
+            if (test not in instrumented_test_result):
+                print(f"Test {test} not found in instrumented test results")
+                failed = True
+                break
+            if test == invariance_test_node_id:
+                if (result == instrumented_test_result[test]):
+                    print(f"Test {test} results match for invariance test")
+                    failed = True
+                    break
+                else:
+                    continue
+            if (result != instrumented_test_result[test]):
+                print(f"Test {test} results do not match")
+                failed = True
+                break
+        print("All tests passed for project", project.name)
+        sys.path.remove(str(project / "src"))
+        sys.path.remove(str(project.parent))
+
+    if failed:
+        print("Test results do not match before and after instrumentation")
+        return False
+    
+    return True
+
+
+def install_project(projects_dir):
+    subprocess.run(["pip", "install", "-e", str(projects_dir)])
+    requirements_file = projects_dir / "requirements.txt"
+    if requirements_file.exists():
+        subprocess.run(["pip", "install", "-r", str(requirements_file)])
+    sys.path.append(str(projects_dir / "src"))
+    sys.path.append(str(projects_dir.parent))
+
+
+def run_test_on_projects():
+    is_successful = True
+    dynapyt_dir = Path(__file__).parent.parent
+    execution_invariance_test_dir = Path(__file__).parent
+    test_folders_file = execution_invariance_test_dir / "test_folders.txt"
+    projects = []
+    project_module_map = {}
+    with open(test_folders_file, "r") as f:
+        while True:
+            line = f.readline()
+            if not line:
+                break
+            project_dir, project_module = line.strip().split()
+            project = dynapyt_dir / project_dir
+            projects.append(project)
+            project_module_map[project.name] = project_module
+
+    if run_tests(projects, project_module_map):
+        print("All tests passed for all projects")
+    else:
+        print("Some tests failed")
+
+    if not is_successful:
+        raise ValueError("Some tests failed")
```

Comment:
Why `ValueError`? There should be a more appropriate exception type. It might also be okay to use exit codes.